src/tactic/library_search.lean

```diff
@@ -76,29 +76,62 @@ meta def library_defs (hs : name) : tactic (list decl_data) :=
 do env ← get_env,
    return $ env.decl_filter_map (process_declaration hs)
 
-meta def apply_and_solve (discharger : tactic unit) (e : expr) :=
+/-- Fail if the target contains a metavariable. -/
+meta def no_mvars_in_target : tactic unit :=
+do t ← target,
+   guard t.list_meta_vars.empty
+
+/--
+Apply the lemma `e`, then attempt to close all goals using `solve_by_elim { discharger := discharger }`,
+failing if `close_goals = tt` and there are any goals remaining.
+-/
+-- Implementation note: as this is used by both `library_search` and `suggest`,
+-- we first run `solve_by_elim` separately on a subset of the goals, whether or not `close_goals` is set,
+-- and then if `close_goals = tt`, require that `solve_by_elim { all_goals := tt }` succeeds
+-- on the remaining goals.
+meta def apply_and_solve (close_goals : bool) (discharger : tactic unit) (e : expr) : tactic unit :=
```

Comment:
I'll add some further explanation, but let me try here first.

`solve_by_elim` is happy to solve a non-singleton goal using a local hypothesis (e.g. fill in a `nat`), even if some other goal depends on it. Thus calling `solve_by_elim` separately on each goal can cause it to fill in a bad solution to one goal, without the possibility of backtracking later.
The mode `solve_by_elim { all_goals := tt }` performs backtracking across multiple goals. However, `solve_by_elim` fails unless it discharges all the goals it's been given.

What behaviour do we want? For `suggest`, we would like "as many subgoals as possible" discharged for us by `solve_by_elim` after the application of a library lemma. The current code attempts to first discharge any "safe" goals, i.e. purely propositional goals not containing metavariables. (Other goals can't depend on the solution we pick for a propositional goal, by proof irrelevance, unless perhaps the goal involves a metavariable. As an example, if you try to prove `a = b` by apply `eq.trans`, the first goal is `a = ?m`, which looks propositional so `solve_by_elim` would happily solve using `rfl`, which is likely a mistake.)

After we've attempted discharging the "safe" goals, we then do a second pass and try to solve all the goals together (with backtracking). If `close_goals = tt`, we insist that this second pass succeeds. This step allows us to fill in "data" solutions for goals, as long as we subsequently solve all the goals.

There is a potential performance problem here: the "safe" goals will (likely) be considered _twice_ by `solve_by_elim`. (In fact, if there are unsafe goals, in the second pass `solve_by_elim` will work on those first, just because they are typically first in the list of goals, and not actually get to the safe goals unless it has an idea for the unsafe goals.) However if we only have "safe" goals, but we can't solve them, the current implementation does try them twice, which is pretty unfortunate.

I *think* a satisfactory solution to this may be to check "does there exist a non-safe goal", and only run the second phase of `solve_by_elim` if this is the case. This should remove nearly all the duplication of effort.

Reply:
Adding this test seems to work well. I've committed an update, that also includes a much more detailed discussion of the logic in `apply_and_solve`.

Reply:
Okay, thanks for clarifying, this makes much more sense now. I had assumed `solve_by_elim {all_goals := tt}` was the same as `all_goals {solve_by_elim}`, but that's not the case. This naming is slightly confusing.

Reply:
Regarding my first comment: I noticed `library_search` fails immediately unless there's exactly one goal, so `apply_and_solve` should never get multiple goals to start with, right? Then my comment doesn't matter for that case. But I think it doesn't hurt to put `focus1 $` at the very beginning of `apply_and_solve` just in case it ends up being used somewhere else. (If not, maybe note in the doc string that it assumes there's only one goal.)

Reply:
I think I'll just fix `library_search` and `suggest` so that they don't mind being given multiple goals (by using `focus1`).

Reply:
I wonder if there is a clearer name for the `all_goals` option in `solve_by_elim`? `backtrack_all_goals`, perhaps? Long-winded, but much clearer intent.